### DISCUSSION TITLE
magick-wasm exports invalid module-type file

### DISCUSSION THREAD

**Discussion by iambumblehead** [dlemstra/magick-wasm#119] [General]
### magick-wasm version

0.0.24

### Description

```bash  
ReferenceError [Error]: require is not defined
      at file:///[...]/@imagemagick/magick-wasm/dist/index.mjs:22:18
      at file:///[...]/@imagemagick/magick-wasm/dist/index.mjs:7298:7
```

index.mjs has this...
```javascript
var Ut = require("fs"), jt = require("path");
pe ? ae = jt.dirname(ae) + "/" : ae = __dirname + "/", et = (t, r) => (t = St(t) ? new URL(t) : jt.normalize(t), Ut.readFileSync(t, r ? void 0 : "utf8")), dt = (t) => {
```

### Steps to Reproduce

1. import then initialize magick-wasm from within a module-type file `initializeImageMagick('@imagemagick/magick-wasm')`
2. see error `ReferenceError [Error]: require is not defined`

**Comment by dlemstra** [comment]
I don't understand how you are using this project. This error must be related to how you are using this project.

**Reply by iambumblehead** [reply]
The response is confusing. Is the mjs file usable when it is imported to another package? No it is not, because it calls require which is defined in commonjs runtimes and is not defined in module runtimes.

`ReferenceError [Error]: require is not defined` this is an uncontrolled runtime error aka bug. Moving the bug report to a discussion implies there is no bug. What is the reason for doing that?

related https://nodejs.org/api/esm.html#modules-ecmascript-modules

**Reply by iambumblehead** [reply]
https://www.npmjs.com/package/@imagemagick/magick-wasm?activeTab=code

open /@imagemagick/magick-wasm/dist/index.mjs and search "require" and one will find many places that call require

**Reply by iambumblehead** [reply]
in a module runtime `require` needs to be explicitly created and defined like this,
```javascript
import module from 'module'

const require = module.createRequire(import.meta.url)
```
using `require` that way inside a module environment is unusual and `import` is commonly used instead

**Reply by dlemstra** [reply]
I am using `vite build` to create the `dist` files so I have no control on how the output is created. This feels like something you should report upstream? It could also be caused by me using `esnext` as the `target`?

**Reply by iambumblehead** [reply]
if possible, I would recommend _not_ exporting the mjs file and, instead, only export the commonjs version.

When magick-wasm is imported by a module-type package, it will be forced to load the commonjs file and node's runtime will switch over to commnjs for magick-wasm's import tree.

What you want is something like this
```diff
    "exports": {
        ".": {
-           "import": "./dist/index.mjs",
+           "default": "./dist/index.umd.js",
            "require": "./dist/index.umd.js",
            "types": "./dist/index.d.ts"
        },
        "./magick.wasm": "./dist/magick.wasm"
    },
```


> This feels like something you should report upstream?

Politely, magick-wasm exports an invalid file. Maybe it is caused by vite build but, regardless; magick-wasm exports an invalid file. The primary responsible party for the files exported by magick-wasm is magick-wasm, not vite

**Reply by dlemstra** [reply]
Your patch feels incorrect since the `./dist/index.mjs` no longer seems to be used. Maybe @Peeterush can shed some light on this since he made these changes to the build setup.

**Reply by iambumblehead** [reply]
I have contributed to this "discussion" (which should be a bug) in a helpful way. Your evasive and grouchy responses aren't great.

**Reply by dlemstra** [reply]
You consider this a bug but I would first like to hear from Peeterush why they set it up this way. And maybe you should read this discussion again to check who is writing grouchy responses?

**Reply by Peeterush** [reply]
I think most modern tools can overcome having `require()` in an ESM, which is probably why nobody ran into this before. But it is not right indeed.

Just giving up on ESM and removing it out right is not a very good approach of course, so I've done some more research.

The `require()` calls come from the Emscripten output of [`@dlemstra/magick-native`](https://github.com/dlemstra/Magick.Native). I've set up up Vite to ignore them, following the example of [`Armster15/imagemagick-wasm-builds`](https://github.com/Armster15/imagemagick-wasm-builds). But now it turns out that that just keeps the `require()` calls in the ESM output.

While playing around with some build options to solve this, I found that Vite 4 (and up) plays differently both with build in modules as well as `ws` (also consuming; `ws` now needs to be present in your own project). I found that _not_ ignoring them still results in a functional build with Vite 4, and properly gets rid of the `require()` statements!

In #121 I propose just that: not to ignore the dependencies in question. This should solve both the `require()` in ESM issue, as well as the need to install `ws` into your own project when using Vite 4.

**Reply by dlemstra** [reply]
Thanks for the research and PR @Peeterush. I will merge this and publish a new release that hopefully resolves @iambumblehead their issues.

**Comment by dlemstra** [comment]
Do you need a new release to test this @iambumblehead or do you have other methods to test this?

**Reply by iambumblehead** [reply]
@dlemstra I could verify the change if you did make a release.

I recommend, however, adding one unit-test which uses the import version of magick-wasm and verifies that it does work. I don't have bandwidth to submit a PR that would the test myself, but that would be the best way to verify, imo.

If you make a release, I'll try using it and will reply here soon and share the results with you and @Peeterush

**Reply by iambumblehead** [reply]
_In the event you may want a quick reply nowish, I may be afk for the next hour or two_

**Reply by dlemstra** [reply]
I will publish the release once the `connectedComponents` PR has been merged. But I will first give the user some some time to work on this instead of merging it and making the changes myself.

**Reply by iambumblehead** [reply]
@dlemstra thanks for letting me know the plan. I will be ready try it out whenever it may become ready in the future.

**Reply by dlemstra** [reply]
Just realized that you could do an `npm run build` and copy the output from the `dist` folder to the `node_modules`  folder of a project where you use this library?

**Reply by iambumblehead** [reply]
@nekowinston did build the latest sources and encountered new errors https://github.com/dlemstra/magick-wasm/discussions/119#discussioncomment-7472770

**Reply by nekowinston** [reply]
Yeah, I built it locally with `npm ci && npm run build && npm pack`, then simply did `yarn add -D path_to.tgz` in my other project. NPM should be able to also add local paths to package.json.

**Comment by nekowinston** [comment]
I just ran into a similar issue when trying to use `import` with [`tsx`](https://github.com/esbuild-kit/tsx):
```console
file:///Users/winston/Code/catppuccin/vscode-icons/node_modules/@imagemagick/magick-wasm/dist/index.mjs:22
        var Ut = require("fs"), jt = require("path");
                 ^

ReferenceError: require is not defined in ES module scope, you can use import instead
    at <anonymous> (/Users/winston/Code/catppuccin/vscode-icons/node_modules/@imagemagick/magick-wasm/dist/index.mjs:8072:980)
    at <anonymous> (/Users/winston/Code/catppuccin/vscode-icons/node_modules/@imagemagick/magick-wasm/dist/index.mjs:8072:980)
    at new Promise (<anonymous>)
    at f.loader (/Users/winston/Code/catppuccin/vscode-icons/node_modules/@imagemagick/magick-wasm/dist/index.mjs:8072:980)
    at f._initialize (/Users/winston/Code/catppuccin/vscode-icons/node_modules/@imagemagick/magick-wasm/dist/index.mjs:8072:980)
    at Qu (/Users/winston/Code/catppuccin/vscode-icons/node_modules/@imagemagick/magick-wasm/dist/index.mjs:8072:980)
    at <anonymous> (/Users/winston/Code/catppuccin/vscode-icons/scripts/catwalk.ts:14:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:194:25)

Node.js v18.18.2
```

I pulled the latest `main` with #121 applied, built it myself, now I'm getting past the import/require, but I run into `__dirname` being not defined:
```console
file:///Users/winston/Code/catppuccin/vscode-icons/node_modules/@imagemagick/magick-wasm/dist/index.mjs:56
        pe ? ie = $t.dirname(ie) + "/" : ie = __dirname + "/", et = (t, r) => (t = St(t) ? new URL(t) : $t.normalize(t), jt.readFileSync(t, r ? void 0 : "utf8")), dt = (t) => {
                                              ^

ReferenceError: __dirname is not defined in ES module scope
    at file:///Users/winston/Code/catppuccin/vscode-icons/node_modules/@imagemagick/magick-wasm/dist/index.mjs:56:47
    at file:///Users/winston/Code/catppuccin/vscode-icons/node_modules/@imagemagick/magick-wasm/dist/index.mjs:7347:7
    at new Promise (<anonymous>)
    at f.loader (file:///Users/winston/Code/catppuccin/vscode-icons/node_modules/@imagemagick/magick-wasm/dist/index.mjs:7341:26)
    at f._initialize (file:///Users/winston/Code/catppuccin/vscode-icons/node_modules/@imagemagick/magick-wasm/dist/index.mjs:7363:16)
    at no (file:///Users/winston/Code/catppuccin/vscode-icons/node_modules/@imagemagick/magick-wasm/dist/index.mjs:7400:12)
    at <anonymous> (/Users/winston/Code/catppuccin/vscode-icons/scripts/catwalk.ts:14:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:194:25)

Node.js v18.18.2
```

Code to reproduce:
`catwalk.ts`
```ts
import { initializeImageMagick } from "@imagemagick/magick-wasm";
await initializeImageMagick();
```

`package.json`, with a local path@ e679d6d
```json
{
  "type": "module",
  "devDependencies": {
    "@imagemagick/magick-wasm": "/Users/winston/Code/magick-wasm/imagemagick-magick-wasm-0.0.24.tgz",
    "tsx": "^3.12.7"
  }
}
```

```console
$ tsx ./catwalk.ts
> results in error above
```

**Reply by iambumblehead** [reply]
its a little disappointing

**Reply by nekowinston** [reply]
Also worth noting that `npm run demo` doesn't work for me @ e679d6db737f76f849f1b623670ee65f9a1a29f4:
https://gist.github.com/nekowinston/d753403bb8465e13308a7eea5f641a43#file-gistfile0-txt-L51

**Reply by dlemstra** [reply]
It looks like those recent changes from @Peeterush broke the demo. I will revert that change for now and add demo to the build. I suspect the problem is that the `magick-native` library creates a `.js` file instead of both a `.js` and `.mjs` file. And thanks for providing a sample and instructions on how I can reproduce this. Makes it a lot easier for me to try to figure out what is going on.

**Reply by Peeterush** [reply]
I've made a new pull request (#123), using [`@rollup/plugin-esm-shim`](https://github.com/rollup/plugins/tree/master/packages/esm-shim). But indeed, if `magick-native` can export its code as ESM natively, there's no more need for back and forth conversion.

**Reply by iambumblehead** [reply]
It would be great if at least one hello-world-style unit-test would import the esm-version of magick-wasm. Changes in the previous PR failed because they had not been tested or verified by anyone.

**Reply by dlemstra** [reply]
I just published a new release that includes the patches made by Peeterush.

**Reply by iambumblehead** [reply]
locally verified: magick-wasm can be imported and the bytes file loaded using esm.

I wanted to report back several months ago but things came up and timing was not good.

**Reply by iambumblehead** [reply]
all is working well here. thank you :)
